design/cluster-api-provider-metal3/dhcpless_ipam.md

```diff
@@ -0,0 +1,397 @@
+  a tool such as `cloud-init` which supports the OpenStack network data format.
+* `preprovisioningNetworkDataName` - this data is designed to allow passing data
+  during the preprovisioning phase, e.g to configure networking for the IPA deploy
+  ramdisk.
+
+The `preprovisioningNetworkDataName` API was added initially to enable [image
+building workflows](https://github.com/metal3-io/baremetal-operator/blob/main/docs/api.md#preprovisioningimage), and a [recent BMO change](https://github.com/metal3-io/baremetal-operator/pull/1380) landed to enable this flow without any custom PreprovisioningImage controller.
+
+#### IPAM configuration
+
+As a user I wish to make use use of the Metal<sup>3</sup> IPAM solution, in a
+DHCP-less environment.
+
+Metal<sup>3</sup> provides an [IPAM controller](https://github.com/metal3-io/ip-address-manager)
+which can be used to allocate IPs used as part of the Metal3Machine lifecycle.
+
+Some gaps exist which prevent realizing this flow in a fully DHCP-less environment,
+so the main focus of the proposal will be how to solve for this use-case.
+
+##### IPAM Scenario 1 - common IPPool
+
+An environment where a common configuration is desired for the pre-provisionining
+phase and the provisioned BareMetalHost (e.g scenario where hosts are permanentaly
+assigned to specific clusters)
+
+##### IPAM Scenario 2 - decoupled preprovisioning/provisioning IPPool
+
+An environment where a decoupled configuration is desired for the pre-provisionining
+phase and the provisioned BareMetalHost (e.g BMaaS scenario where end-user network configuration
+differs from the commissioning phase where a different configuration is desired for inspection/cleaning)
+
+## Design Details
+
+`Metal3MachineTemplate` and `Metal3DataTemplate` are used to apply networkData to specific BareMetalHost resources,
+but they are by design coupled to the CAPI Machine lifecycle.
+
+This is a problem for the pre-provisioning use-case since at this point we're preparing the BareMetalHost for
+use, there is not yet any Machine.
+
+To resolve this below we outline a proposal to add two new resources with similar behavior for the pre-provisioning
+phase `Metal3PreProvisioningTemplate` and `Metal3PreProvisioningDataTemplate`
+
+### API overview
+
+The current flow in the provisioning phase is as follows (only the most relevant fields are included for clarity):
+
+```yaml
+apiVersion: ipam.metal3.io/v1alpha1
+kind: IPPool
+metadata:
+  name: pool-1
+spec:
+  clusterName: cluster
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3DataTemplate
+metadata:
+  name: data-template
+spec:
+  clusterName: cluster
+  networkData:
+    networks:
+      ipv4:
+      - id: eth0
+        ipAddressFromIPPool: pool-1
+
+---
+
+apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
+kind: Metal3MachineTemplate
+metadata:
+  name: machine-template
+spec:
+  template:
+    spec:
+      dataTemplate:
```

Comment:
It feels like you need to add a link to the Meta3PreprovisioningDataTemplate here rather than adding another template object?

Reply:
Hmm, it's probably too late to do it here, never mind.

Reply:
Yes at the point of consuming the preprovisioningNetworkData the BMH is not associated with any Machine yet so we need another way to consume the template resource and link it to the BMH.